### PR TITLE
fix(ivy): Support selector-less directive as base classes

### PR DESCRIPTION
**Pull request by atscott**
Following #31379, this adds support for directives without a selector to
Ivy.

## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?

- [x] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] angular.io application / infrastructure changes
- [ ] Other... Please describe:

## What is the current behavior?
Only VE supports selector-less directives, per #31379.

Issue Number: N/A

## What is the new behavior?
Ivy also supports selector-less directives

## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No

### CODE REVIEW

File: packages/compiler-cli/src/ngtsc/annotations/src/ng_module.ts [hunk 1]
```diff
@@ -210,15 +211,26 @@ export class NgModuleDecoratorHandler implements DecoratorHandler<NgModuleAnalys
...
           analysis.ngInjectorDef.imports.push(this.refEmitter.emit(exportRef, context));
         }
       }
+
+      for (const decl of analysis.declarations) {
+        // All pipes are registered but only directives with selectors are. If it is neither a pipe
+        // nor a registered directive, then it is a directive without a selector so throw an error
+        // because directives in declarations list must have a selector.
+        if (!this.metaReader.getPipeMetadata(decl) && !this.metaReader.getDirectiveMetadata(decl)) {
+          throw new FatalDiagnosticError(
```
Comment (alxhub) [inline packages/compiler-cli/src/ngtsc/annotations/src/ng_module.ts]: Another thing that can cause this is that they added a class to `declarations` which doesn't have any annotations, so I don't think this error message is correct in all cases.

Instead, we should probably register selectorless directives with the `MetadataRegistry` so we can explicitly query about it here.

**Review by alxhub — CHANGES_REQUESTED** [review]

File: packages/compiler-cli/src/ngtsc/annotations/src/directive.ts [hunk 2]
```diff
@@ -112,17 +112,22 @@ export function extractDirectiveMetadata(
   metadata: R3DirectiveMetadata,
   decoratedElements: ClassMember[],
 }|undefined {
-  if (decorator.args === null || decorator.args.length !== 1) {
+  let directive: Map<string, ts.Expression>;
+  if (decorator.args === null || decorator.args.length === 0) {
```
Comment (AndrewKushnir) [inline packages/compiler-cli/src/ngtsc/annotations/src/directive.ts]: I'd propose to add a quick comment here saying that we allow `Directive` decorator to be used without arguments for abstract directive cases, but we do additional verification at NgModule level.

File: packages/core/test/linker/integration_spec.ts [hunk 3]
```diff
@@ -1425,6 +1425,21 @@ function declareTests(config?: {useJit: boolean}) {
             .toThrowError(`Directive ${stringify(SomeDirective)} has no selector, please add it!`);
       });
 
+      it('should not throw when using directives without selector as base class not in declarations',
+         () => {
+           @Directive({})
```
Comment (AndrewKushnir) [inline packages/core/test/linker/integration_spec.ts]: Let's may be extend this test with additional cases that we have in `ngtsc_spec.ts` (added in this PR)?

**Review by AndrewKushnir — CHANGES_REQUESTED** [review]
Looks great, just a couple nits. Thanks @atscott 👍

File: packages/core/test/linker/integration_spec.ts (hunk 3 above)
Comment (atscott) [reply packages/core/test/linker/integration_spec.ts]: Done.

File: packages/compiler-cli/src/ngtsc/annotations/src/directive.ts (hunk 2 above)
Comment (atscott) [reply packages/compiler-cli/src/ngtsc/annotations/src/directive.ts]: Done.

**Review by AndrewKushnir — APPROVED** [review]
LGTM 👍

File: packages/compiler-cli/src/ngtsc/annotations/src/ng_module.ts (hunk 1 above)
Comment (atscott) [reply packages/compiler-cli/src/ngtsc/annotations/src/ng_module.ts]: Done. PTAL

File: packages/compiler-cli/src/ngtsc/annotations/src/ng_module.ts [hunk 4]
```diff
@@ -43,7 +43,8 @@ export class NgModuleDecoratorHandler implements DecoratorHandler<NgModuleAnalys
       private metaRegistry: MetadataRegistry, private scopeRegistry: LocalModuleScopeRegistry,
       private referencesRegistry: ReferencesRegistry, private isCore: boolean,
       private routeAnalyzer: NgModuleRouteAnalyzer|null, private refEmitter: ReferenceEmitter,
-      private defaultImportRecorder: DefaultImportRecorder, private localeId?: string) {}
+      private defaultImportRecorder: DefaultImportRecorder, private metaReader: MetadataReader,
```
Comment (alxhub) [inline packages/compiler-cli/src/ngtsc/annotations/src/ng_module.ts]: Can you put the `metaReader` argument before the `metaRegistry`? This mirrors some other code which takes both arguments.

File: packages/core/testing/src/r3_test_bed_compiler.ts [hunk 5]
```diff
@@ -299,7 +299,7 @@ export class R3TestBedCompiler {
     this.pendingComponents.clear();
 
     this.pendingDirectives.forEach(declaration => {
-      const metadata = this.resolvers.directive.resolve(declaration) !;
+      const metadata = this.resolvers.directive.resolve(declaration);
```
Comment (alxhub) [inline packages/core/testing/src/r3_test_bed_compiler.ts]: +1 on the change, but this seems unrelated. Did our typings change in a way that made this no longer return `null`?

File: packages/compiler-cli/src/ngtsc/metadata/src/dts.ts [hunk 6]
```diff
@@ -21,6 +21,11 @@ import {extractDirectiveGuards, extractReferencesFromType, readStringArrayType,
 export class DtsMetadataReader implements MetadataReader {
   constructor(private checker: ts.TypeChecker, private reflector: ReflectionHost) {}
 
+  isAbstractDirective(ref: Reference<ClassDeclaration>): boolean {
+    const meta = this.getDirectiveMetadata(ref);
+    return meta ? (meta.selector === null || meta.selector === '') : false;
```
Comment (alxhub) [inline packages/compiler-cli/src/ngtsc/metadata/src/dts.ts]: Explicitly check for `null`ness of `meta`, don't rely on truthiness.

I think we should treat `selector: ''` as an error and not allow it to be produced in the first place. Thoughts?

File: packages/compiler-cli/src/ngtsc/annotations/src/directive.ts (hunk 2 above)
Comment (alxhub) [reply packages/compiler-cli/src/ngtsc/annotations/src/directive.ts]: I think the comment should go in place here, and not on the jsdoc of the function (since it's somewhat unrelated to the purpose of `extractDirectiveMetadata`.

File: packages/compiler-cli/src/ngtsc/annotations/src/directive.ts [hunk 7]
```diff
@@ -188,9 +200,9 @@ export function extractDirectiveMetadata(
     }
     // use default selector in case selector is an empty string
     selector = resolved === '' ? defaultSelector : resolved;
-  }
-  if (!selector) {
-    throw new Error(`Directive ${clazz.name.text} has no selector, please add it!`);
+    if (!selector) {
+      throw new Error(`Directive ${clazz.name.text} has no selector, please add it!`);
```
Comment (alxhub) [inline packages/compiler-cli/src/ngtsc/annotations/src/directive.ts]: Let's go ahead and convert this to your `DIRECTIVE_MISSING_SELECTOR` diagnostic.

**Review by alxhub — CHANGES_REQUESTED** [review]

File: packages/compiler-cli/src/ngtsc/annotations/src/directive.ts (hunk 7 above)
Comment (atscott) [reply packages/compiler-cli/src/ngtsc/annotations/src/directive.ts]: Done

File: packages/compiler-cli/src/ngtsc/annotations/src/ng_module.ts (hunk 4 above)
Comment (atscott) [reply packages/compiler-cli/src/ngtsc/annotations/src/ng_module.ts]: Done

File: packages/compiler-cli/src/ngtsc/metadata/src/dts.ts (hunk 6 above)
Comment (atscott) [reply packages/compiler-cli/src/ngtsc/metadata/src/dts.ts]: Agreed. We actually do throw errors elsewhere if you explicitly set `selector: ''`

File: packages/core/testing/src/r3_test_bed_compiler.ts (hunk 5 above)
Comment (atscott) [reply packages/core/testing/src/r3_test_bed_compiler.ts]: I think the difference is that this can *actually* return `null` now whereas before it never would. Since we allow `@Directive()`, this would return `null` but before we required an object and this call would always(?) return something. The `metadata` parameter of `compileDirective` call two lines below now allows `null` so I felt this was more accurate to what's possible instead of using the null assertion operator.

**Review by alxhub — APPROVED** [review]

**Comment by atscott** [comment]
[Ivy presubmit](https://test.corp.google.com/ui#id=OCL:263595893:BASE:264258073:1566254227063:a39433f5)
[VE presubmit](https://test.corp.google.com/ui#id=OCL:263595893:BASE:264257690:1566254122897:49c5c67c)
